source/common/common/recursion_checker.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Common {
+/**
+ * A helper class to prevent re-entry into a non-reentrant class.
+ */
+class RecursionChecker {
+public:
+  void enter() {
+    RELEASE_ASSERT(!entered_, "A resource should only be entered once");
```

Comment:
I would make this a normal `ASSERT` as this is a programming error that shouldn't happen. I suspect also that if you guard the member variable access with !NDEBUG, you can get all of this to compile completely away in production builds? Then maybe rename `DebugRecursionChecker` or similar?